OpenScienceJournal/whistlepunk_library/src/main/java/com/google/android/apps/forscience/whistlepunk/devicemanager/NativeBleDiscoverer.java

```diff
@@ -41,19 +70,28 @@
     private static final SensorProvider PROVIDER = new SensorProvider() {
         @Override
         public SensorChoice buildSensor(String sensorId, ExternalSensorSpec spec) {
-            return new BluetoothSensor(sensorId, (BleSensorSpec) spec,
-                    BluetoothSensor.ANNING_SERVICE_SPEC);
+            if (spec instanceof MkrSciBleSensorSpec) {
+                return new MkrSciBleSensor(sensorId, (MkrSciBleSensorSpec) spec);
+            } else {
+                return new BluetoothSensor(sensorId, (BleSensorSpec) spec,
+                        BluetoothSensor.ANNING_SERVICE_SPEC);
+            }
         }
 
         @Override
         public ExternalSensorSpec buildSensorSpec(String name, byte[] config) {
-            return new BleSensorSpec(name, config);
+            try {
+                return new MkrSciBleSensorSpec(name, config);
```

Comment:
When I run the SensorRegistryTest.testLoggingId test, it fails because this code here is able to create a MkrSciBleSensorSpec from a config when it should have created a BleSensorSpec. After the MkrSciBleSensorSpec is created, it is passed to the MkrSciBleSensor constructor (see line 74 above) and that throws an exception with the message "Unmanaged mkr sci ble sensor: 3".

Reply:
To fix this, I've added a method to validate the spec.
`  public static void validateSpec(MkrSciBleSensorSpec spec) {
    String sensorKind = spec.getSensor();
    switch (sensorKind) {
      case SENSOR_INPUT_1:
      case SENSOR_INPUT_2:
      case SENSOR_INPUT_3:
      case SENSOR_VOLTAGE:
      case SENSOR_CURRENT:
      case SENSOR_RESISTANCE:
      case SENSOR_ACCELEROMETER_X:
      case SENSOR_ACCELEROMETER_Y:
      case SENSOR_ACCELEROMETER_Z:
      case SENSOR_LINEAR_ACCELEROMETER:
      case SENSOR_GYROSCOPE_X:
      case SENSOR_GYROSCOPE_Y:
      case SENSOR_GYROSCOPE_Z:
      case SENSOR_MAGNETOMETER:
        break;
      default:
        throw new RuntimeException("Unmanaged mkr sci ble sensor: " + sensorKind);
    }
  }
`